File: bank-scraper/application/ports/position_port.py
import abc
from datetime import datetime
from typing import Optional

from domain.financial_entity import FinancialEntity
from domain.global_position import GlobalPosition


class PositionPort(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def save(self, position: GlobalPosition):
        raise NotImplementedError

    @abc.abstractmethod
    def get_last_grouped_by_entity(self) -> dict[FinancialEntity, GlobalPosition]:
        raise NotImplementedError

    @abc.abstractmethod
    def get_last_updated(self, entity_id: int) -> Optional[datetime]:
        raise NotImplementedError

File: mobile/src/application/ports/core/positionPort.ts
import {
  DataSource,
  Entity,
  FundDetail,
  GlobalPosition,
  PositionQueryRequest,
  ProductType,
  StockDetail,
} from "@/domain"
import { Dezimal } from "@/domain/dezimal"

export interface PositionPort {
  save(position: GlobalPosition): Promise<void>
  getLastGroupedByEntity(
    query?: PositionQueryRequest | null,
  ): Promise<Map<Entity, GlobalPosition>>
  deletePositionForDate(
    entityId: string,
    date: string,
    source: DataSource,
  ): Promise<void>
  getById(positionId: string): Promise<GlobalPosition | null>
  deleteById(positionId: string): Promise<void>
  getStockDetail(entryId: string): Promise<StockDetail | null>
  getFundDetail(entryId: string): Promise<FundDetail | null>
  updateMarketValue(
    entryId: string,
    productType: ProductType,
    marketValue: Dezimal,
  ): Promise<void>
}
